src/dune_engine/scheduler.ml

```diff
@@ -643,8 +676,101 @@ module Handler = struct
   type t = Config.t -> Event.t -> unit
 end
 
+module Timer : sig
+  type t
+
+  val create : Event.Queue.t -> frequency:float -> t
+
+  type sleeper
+
+  val await : sleeper -> [ `Finished | `Cancelled ] Fiber.t
+
+  val cancel : t -> sleeper -> unit Fiber.t
+
+  val sleep : t -> float -> sleeper
+
+  val close : t -> unit
+end = struct
+  type sleeper = [ `Finished | `Cancelled ] Fiber.Ivar.t
+
+  type t =
+    { events : Event.Queue.t
+    ; mutex : Mutex.t
+    ; frequency : float
+    ; mutable sleepers : (float * [ `Finished | `Cancelled ] Fiber.Ivar.t) list
+    ; mutable active : bool
+    }
+
+  let await = Fiber.Ivar.read
+
+  let cancel t sleeper =
+    Mutex.lock t.mutex;
+    t.sleepers <-
+      List.filter t.sleepers ~f:(fun (_, sleeper') -> not (sleeper' == sleeper));
+    Mutex.unlock t.mutex;
+    let* res = Fiber.Ivar.peek sleeper in
+    match res with
+    | None -> Fiber.Ivar.fill sleeper `Cancelled
+    | Some _ -> Fiber.return ()
+
+  let polling_loop t () =
+    let rec loop () =
+      Mutex.lock t.mutex;
+      match t.active with
+      | false -> ()
+      | true ->
+        let now = Unix.gettimeofday () in
+        t.sleepers <-
+          List.filter_map t.sleepers ~f:(fun (expiration, ivar) ->
+              if now < expiration then
+                Some (expiration, ivar)
+              else (
+                Event.Queue.send_timer_completed t.events
```

Comment:
We're taking a mutex here while holding another mutex, which is deadlock-prone unless done with care.
By doing this we rely on lock ordering and we should probably write that down somewhere. (the documentation of the [mutex] field?)

Reply:
Not a great excuse, but we already do this in many other places in the scheduler (see process table for example). If we decide to document this, we should do it exhaustively.

Reply:
Well, it's never too late to start.

Reply:
In fact the CI is failing with "Mutex.lock: Resource deadlock avoided". Suspicious.

Reply:
Indeed. In my rebasing, I accidentally removed a `Mutex.unlock` in cancel. this is now fixed.

By the way, I updated the code to never acquire nested locks. So the comment isn't event necessary any longer.